FWIO/RNTuple/plugins/RNTupleOutputFile.cc

```diff
@@ -0,0 +1,583 @@
+      model->AddField(std::move(field));
+    }
+    {
+      auto field =
+          ROOT::Experimental::RFieldBase::Create("BranchIDLists", "std::vector<std::vector<unsigned int> >").Unwrap();
+      model->AddField(std::move(field));
+    }
+    {
+      auto field =
+          ROOT::Experimental::RFieldBase::Create("ThinnedAssociationsHelper", "edm::ThinnedAssociationsHelper").Unwrap();
+      model->AddField(std::move(field));
+    }
+    {
+      auto field = ROOT::Experimental::RFieldBase::Create("ProductDependencies", "edm::BranchChildren").Unwrap();
+      model->AddField(std::move(field));
+    }
+
+    auto writeOptions = ROOT::Experimental::RNTupleWriteOptions();
+    writeOptions.SetCompression(convert(iConfig.compressionAlgo), iConfig.compressionLevel);
+    metaData_ = ROOT::Experimental::RNTupleWriter::Append(std::move(model), "MetaData", file_, writeOptions);
+  }
+
+  void RNTupleOutputFile::fillMetaData(BranchIDLists const& iBranchIDLists,
+                                       ThinnedAssociationsHelper const& iThinnedHelper,
+                                       ProductRegistry const& iReg) {
+    auto rentry = metaData_->CreateEntry();
+
+    FileID id(createGlobalIdentifier());
+
+    rentry->BindRawPtr("FileIdentifier", &id);
+
+    indexIntoFile_.sortVector_Run_Or_Lumi_Entries();
+    rentry->BindRawPtr("IndexIntoFile", &indexIntoFile_);
+
+    ProcessHistoryVector procHistoryVector;
+    for (auto const& ph : processHistoryRegistry_) {
+      procHistoryVector.push_back(ph.second);
+    }
+    rentry->BindRawPtr("ProcessHistory", &procHistoryVector);
+
+    // Make a local copy of the ProductRegistry, removing any transient or pruned products.
+    using ProductList = ProductRegistry::ProductList;
+    ProductRegistry pReg(iReg.productList());
+    ProductList& pList = const_cast<ProductList&>(pReg.productList());
+    for (auto const& prod : pList) {
+      if (prod.second.branchID() != prod.second.originalBranchID()) {
+        if (branchesWithStoredHistory_.find(prod.second.branchID()) != branchesWithStoredHistory_.end()) {
+          branchesWithStoredHistory_.insert(prod.second.originalBranchID());
+        }
+      }
+    }
+    std::set<BranchID>::iterator end = branchesWithStoredHistory_.end();
+    for (ProductList::iterator it = pList.begin(); it != pList.end();) {
+      if (branchesWithStoredHistory_.find(it->second.branchID()) == end) {
+        // avoid invalidating iterator on deletion
+        ProductList::iterator itCopy = it;
+        ++it;
+        pList.erase(itCopy);
+
+      } else {
+        ++it;
+      }
+    }
+
+    rentry->BindRawPtr("ProductRegistry", &pReg);
+    rentry->BindRawPtr("ThinnedAssociationsHelper", const_cast<void*>(static_cast<const void*>(&iThinnedHelper)));
+    rentry->BindRawPtr("BranchIDLists", const_cast<void*>(static_cast<const void*>(&iBranchIDLists)));
+    rentry->BindRawPtr("ProductDependencies", const_cast<void*>(static_cast<const void*>(&productDependencies_)));
```

Comment:
Can these be
```suggestion
    rentry->BindRawPtr("ThinnedAssociationsHelper", static_cast<void*>(&iThinnedHelper));
    rentry->BindRawPtr("BranchIDLists", static_cast<void*>(&iBranchIDLists));
    rentry->BindRawPtr("ProductDependencies", static_cast<const void*>(&productDependencies_));
```
?

Reply:
Since the arguments passed to the function are logically `const` I do not want to change them. The API problem is `BindRawPtr` requiring a non-const `void*` even for the case where we are only reading the data from the pointer.

Reply:
I see. We should then communicate this issue in the API to the ROOT team.

Reply:
@hahnjo @pcanal This kind of code is the reason why we would prefer to have separate `REntry` API for reading and writing.

(we should probably communicate this point again via other channels too)